google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/PubSubApi.scala

```diff
@@ -121,87 +129,138 @@ private[pubsub] trait PubSubApi {
       AcknowledgeRequest(json.asJsObject.fields("ackIds").convertTo[immutable.Seq[String]]: _*)
     def write(ar: AcknowledgeRequest): JsValue = JsObject("ackIds" -> ar.ackIds.toJson)
   }
-  private implicit val pullRequestFormat = DefaultJsonProtocol.jsonFormat2(PubSubApi.PullRequest)
+  private implicit val pullRequestFormat = DefaultJsonProtocol.jsonFormat2(PullRequest)
+
+  private def pool[T]()(
+      implicit as: ActorSystem
+  ): Flow[(HttpRequest, T), (Try[HttpResponse], T), Http.HostConnectionPool] =
+    if (isEmulated) {
+      Http().cachedHostConnectionPool[T](PubSubGoogleApisHost, PubSubGoogleApisPort)
+    } else {
+      Http().cachedHostConnectionPoolHttps[T](PubSubGoogleApisHost, PubSubGoogleApisPort)
+    }
 
-  def pull(project: String,
-           subscription: String,
-           maybeAccessToken: Option[String],
-           returnImmediately: Boolean,
-           maxMessages: Int)(
+  def pull(project: String, subscription: String, returnImmediately: Boolean, maxMessages: Int)(
       implicit as: ActorSystem,
       materializer: Materializer
-  ): Future[PullResponse] = {
+  ): Flow[(Done, Option[String]), PullResponse, NotUsed] = {
     import materializer.executionContext
 
-    val uri: Uri = s"$PubSubGoogleApisHost/v1/projects/$project/subscriptions/$subscription:pull"
-
-    val request = PubSubApi.PullRequest(returnImmediately = returnImmediately, maxMessages = maxMessages)
-
-    for {
-      request <- Marshal((HttpMethods.POST, uri, request)).to[HttpRequest]
-      response <- doRequest(request, maybeAccessToken)
-      pullResponse <- response.status match {
-        case StatusCodes.Success(_) => Unmarshal(response).to[PullResponse]
-        case status =>
-          Unmarshal(response)
-            .to[String]
-            .map { entity =>
-              throw new RuntimeException(s"Unexpected pull response. Code: [$status]. Entity: [$entity]")
-            }
+    val url: Uri = s"/v1/projects/$project/subscriptions/$subscription:pull"
+
+    Flow[(Done, Option[String])]
+      .mapAsync(1) {
+        case (_, maybeAccessToken) =>
+          Marshal((HttpMethods.POST, url, PullRequest(returnImmediately, maxMessages)))
+            .to[HttpRequest]
+            .map(request => (authorize(maybeAccessToken)(request), NotUsed))
+      }
+      .via(pool[NotUsed]())
```

Comment:
Please make add a `pool` variant without context so that the using methods don't get to create strange tuples, such as 
```
  private def poolSimple()(
    implicit as: ActorSystem
  ): Flow[HttpRequest, Try[HttpResponse], Http.HostConnectionPool] =
    Flow[HttpRequest]
      .map(_ -> NotUsed)
      .viaMat(pool())(Keep.right)
      .map(_._1)
```